src/commit-message.ts

```diff
@@ -0,0 +1,45 @@
+import { PullRequestInfo } from './models'
+import { Config } from './config'
+
+export function getCommitMessage (
+  pullRequestInfo: PullRequestInfo,
+  config: Config
+) {
+  return config.mergeCommitMessage
+    ? getCustomCommitMessage(pullRequestInfo, config.mergeCommitMessage)
+    : null
+}
+
+export function splitCommitMessage (message: string) {
+  const [title, ...body] = message.split('\n')
+
+  return {
+    title,
+    body: body.join('\n')
+  }
+}
+
+type Tag = 'title' | 'body' | 'number' | 'branch' | 'commits'
+
+type Tags = {
+  [key in Tag]: (pullRequestInfo: PullRequestInfo) => string
+}
+
+function getCustomCommitMessage (
+  pullRequestInfo: PullRequestInfo,
+  template: string
+) {
+  const tagReplacers: Tags = {
+    title: pullRequestInfo => pullRequestInfo.title,
+    body: pullRequestInfo => pullRequestInfo.body,
+    number: pullRequestInfo => pullRequestInfo.number.toString(),
+    branch: pullRequestInfo => pullRequestInfo.headRefName,
+    commits: pullRequestInfo => pullRequestInfo.allCommits.nodes.map(node => {
+      return `* ${node.commit.messageHeadline} (${node.commit.abbreviatedOid})`
+    }).join('\n\n')
+  }
+
+  return (Object.keys(tagReplacers) as Tag[]).reduce((message, tagName) => {
```

Comment:
I think this should work the other way around to avoid recursive replacements (replacing `{number}` when `{number}` is part of a PRs title or body).

```
return template.replace(/\{(\w+)\}/g, (match, tagName) => {
  const tagReplacer = tagReplacers[tagName]
  if (tagReplacer) {
    return tagReplacer(pullRequestInfo)
  } else {
    return match
  }
})
```

Reply:
Fixed in https://github.com/bobvanderlinden/probot-auto-merge/pull/439/commits/d100f8a06be61b7e3fcba050bc22e818ede5b0a8